nullability/v0.3/nullability-v0.3.graphql

```diff
@@ -0,0 +1,137 @@
+
+`@semanticNonNullField` is the same as `@semanticNonNull` but can be used on type system extensions for services
+that do not own the schema like client services:
+
+```graphql
+# extend the schema to make User.email semantically non-null.
+extend type User @semanticNonNullField(name: "email")
+```
+
+The `levels` argument indicates what levels are semantically non null in case of lists:
+
+```graphql
+# friends is semantically non null
+extend type User @semanticNonNullField(name: "friends")  # same as @semanticNonNullField(name: "friends", levels: [0])
+
+# every friends[k] is semantically non null
+extend type User @semanticNonNullField(name: "friends", levels: [1])
+
+# friends as well as every friends[k] is semantically non null
+extend type User @semanticNonNullField(name: "friends", levels: [0, 1])
+```
+
+`levels` are zero indexed.
+Passing a negative level or a level greater than the list dimension is an error.
+
+See `@semanticNonNull`.
+"""
+directive @semanticNonNullField(name: String!, levels: [Int] = [0]) repeatable on OBJECT | INTERFACE
+
+"""
+Indicates how clients should handle errors on a given position.
+
+When used on the schema definition, `@catch` applies to every position that can return an error.
+
+The `levels` argument indicates where to catch errors in case of lists:
+
+```graphql
+{
+    user {
+        # friends catches errors
+        friends @catch { name } # same as @catch(levels: [0])
+
+        # every friends[k] catches errors
+        friends @catch(levels: [0]) { name }
+
+        # friends as well as every friends[k] catches errors
+        friends @catch(levels: [0, 1]) { name }
+    }
+}
+```
+
+`levels` are zero indexed.
+Passing a negative level or a level greater than the list dimension is an error.
+
+See `CatchTo` for more details.
+"""
+directive @catch(to: CatchTo! = RESULT, levels: [Int] = [0]) on FIELD | FIELD_DEFINITION | SCHEMA
+
+"""
+Indicates how clients should handle errors on a given position.
+
+`@catchField` is the same as `@catch` but can be used on type system extensions for services
+that do not own the schema like client services:
+
+```graphql
+# extend the schema to catch User.email to `RESULT`.
+extend type User @catchField(name: "email", to: RESULT)
+```
+
+See `@catch`.
+"""
+directive @catchField(to: CatchTo! = RESULT, levels: [Int] = [0]) repeatable on INTERFACE | OBJECT
+
+enum CatchTo {
+    """
+    Catch the error and map the position to a result type that can contain either
+    a value or an error.
+    """
+    RESULT,
+    """
+    Catch the error and map the position to a nullable type that will be null
+    in the case of error.
+    This does not allow to distinguish between semantic null and error null but
+    can be simpler in some cases.
+    """
+    NULL,
+    """
+    Throw the error.
+    Parent positions can recover using `RESULT` or `NULL`.
+    If no parent position recovers, the parsing stops.
+    """
+    THROW
```

Comment:
Should we rename this to EXCEPTION while we're at it?

Reply:
I'm a bit hesitant because IIRC relay uses `THROW`. That's something we certainly want to bikeshed and until we do, i'd just keep the current naming to avoid too much back & forth